proxy-setting: Feature to use system proxy settings.

abhigyank opened the pull request:
This commit uses resolveProxy to to read system proxy settings by requesting to a url namely `google.com` using the 4 protocols `http`, `https`, `ftp`, `socks`, this causes the main process to return the system used proxy for this particular protocol. We store this proxies in proper proxy format string using ConfigUtil into settings.json as `systemProxySettings`.
On turning on the use system proxy settings this value of `systemProxySettings`is written into the proxy rules and the webview session and is reloaded as done previously with only manual settings.
It removes the previous use proxy option and replaces it with use system proxy and manual proxy
options.

Fixes: #296
**You have tested this PR on:**
  - [x] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

abhigyank committed:
proxy-setting: Feature to use system proxy settings.

This commit uses resolveProxy to to read system proxy settings and store
them in proper proxy format string using ConfigUtil. It removes the previous
use proxy option and replaces it with use system proxy and manual proxy
options.

Fixes: #296

akashnimare commented on app/main/index.js:
```diff
@@ -151,6 +151,54 @@ app.on('ready', () => {
 
+	const ses = page.session;
+
+	const resolveProxyUrl = 'chat.zulip.org';
```
@abhigyank is this hardcorded?

abhigyank replied:
The main aim is make a request to a website using the mainWindow that can access the system proxy. This `resolveProxyUrl` can be anything, just about any website. Here I have taken `chat.zulip.org` but it could be `google.com` as well.

akashnimare commented on app/main/index.js:
```diff
@@ -151,6 +151,54 @@ app.on('ready', () => {
 	const page = mainWindow.webContents;
 
+	const ses = page.session;
+
```
I think we could put this into a `proxy-util.js`.

abhigyank replied:
Do you mean the entire thing from here till the [end](https://github.com/zulip/zulip-electron/pull/480/files#diff-b7ea71851f2bf6f64dbcec17f174a921R201)? 
I will try to do that. We can use it as `proxyUtil.setSystemProxy()` ?

abhigyank replied:
I  think `storeSystemProxy`/`resolveSystemProxy` sounds appropirate.

abhigyank committed:
proxy-util: Add file proxy-util.js
This file gets system proxy settings and stores in settings.json

abhigyank commented:
@akashnimare Could you also test this on your macOS? If you're not on a proxy server, you can just try putting some random proxy server in system settings and just test if the correct ones are being written on to the `settings.json`.

abhigyank committed:
proxy-setting: Migration from old useProxy to new useManualProxy.

abhigyank committed:
proxy-setting: Use all async resolveProxy function with promises.

abhigyank committed:
proxy-settings: Reload app on changin proxy settings.

abhigyank committed:
fixup: Get system proxy rules only when SPS is checked.

abhigyank committed:
fixup: Remove code for saving proxy rules when turning SPS on.

This is because now we want the user to restart the entire app, so this is no longer needed.

akashnimare commented:
Merged. Thanks, @abhigyank. The only thing remaining here is to update the system proxy settings without restarting the app. The more advance set up would be update the settings etc when user changes/updates these settings from the system configs. Let's also look into `http-proxy` if it gives better results :)
